### PR TITLE
use code splitting to speed up initial page load

### PR DESCRIPTION
**Pull request by sqs**
Uses the [react-router code splitting technique](https://reactjs.org/docs/code-splitting.html#route-based-code-splitting) to reduce the initial bundle size by ~50%. This is intended to make the initial page load faster for users. (It is possible that it makes it slower in some cases because the client now needs to load multiple JS files, possibly in serial. In most cases, we expect there to be a net benefit.)

As a secondary benefit, it speeds up the initial bundle compilation time in local development.

Codemod for one-time migration of static imports to React.lazy imports, change `import` to `ximport` and run:

```
codemod --extensions tsx -d web/src "ximport \{([^}]+)\} from ('[^']+')" 'const \1 = React.lazy(async () => ({ default: (await import(\2)).\1 }))'
````

### CODE REVIEW

File: doc/dev/web_app.md [hunk 1]
```diff
@@ -46,6 +46,22 @@ Their purpose is to reexport symbols from a number of other files to make import
...
+
+``` typescript
+const MyComponent = React.lazy(async () => ({
+    default: (await import('./path/to/MyComponent)).MyComponent,
+}))
+```
+
+If you don't do this (and just use a normal `import`), it will still work, but it will increase the initial bundle size for all users.
+
+(It is necessary to return the component as the `default` property of an object because `React.lazy` is hard-coded to look there for it. We could avoid this extra work by using default exports, but we chose not to use default exports ([reasons](https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad)).)
```
Comment (felixfbecker) [inline doc/dev/web_app.md]: How about we turn on https://palantir.github.io/tslint/rules/match-default-export-name/ instead?

File: web/src/repo/routes.tsx [hunk 2]
```diff
@@ -3,13 +3,17 @@ import { Redirect, RouteComponentProps } from 'react-router'
...
-import { BlobPage } from './blob/BlobPage'
-import { RepositoryCommitsPage } from './commits/RepositoryCommitsPage'
-import { FilePathBreadcrumb } from './FilePathBreadcrumb'
+const BlobPage = React.lazy(async () => ({ default: (await import('./blob/BlobPage')).BlobPage }))
+const RepositoryCommitsPage = React.lazy(async () => ({
+    default: (await import('./commits/RepositoryCommitsPage')).RepositoryCommitsPage,
+}))
+const FilePathBreadcrumb = React.lazy(async () => ({
+    default: (await import('./FilePathBreadcrumb')).FilePathBreadcrumb,
+}))
```
Comment (felixfbecker) [inline web/src/repo/routes.tsx]: I dislike that we are mixing imports and `const` declarations, imo imports should always be at the top (and sorted)

File: web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx [hunk 3]
```diff
@@ -96,22 +81,11 @@ export class DynamicallyImportedMonacoSettingsEditor extends React.PureComponent
...
-                    Error loading: {upperFirst(this.state.monacoSettingsEditorOrError.message)}
-                </div>
-            )
-        }
-
         const isDirty = this.isDirty
         const effectiveValue = this.effectiveValue
 
-        const MonacoSettingsEditor = this.state.monacoSettingsEditorOrError.MonacoSettingsEditor
+        console.log('QQ3')
```
Comment (felixfbecker) [inline web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx]: remove?

File: web/src/repo/routes.tsx (hunk 2 above)
Comment (sqs) [reply web/src/repo/routes.tsx]: Yeah I thought these were sorted but they aren’t. Feel free to fix, or else I will get around to it.

File: web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx (hunk 3 above)
Comment (sqs) [reply web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx]: Already removed

File: web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx [hunk 4]
```diff
@@ -183,8 +159,9 @@ export class DynamicallyImportedMonacoSettingsEditor extends React.PureComponent
 
     private monacoRef = (monacoValue: typeof _monaco | null) => {
         this.monaco = monacoValue
-        // This function can only be called if the editor was loaded so it is okay to cast here
-        const monacoSettingsEditor = this.state.monacoSettingsEditorOrError as typeof _monacoSettingsEditorModule
+        // This function can only be called if the lazy MonacoSettingsEditor component was loaded,
+        // so we know its #_result property is set by now.
+        const monacoSettingsEditor = MonacoSettingsEditor._result
```
[the remark this answers is not in the document]
Comment (sqs) [reply web/src/settings/DynamicallyImportedMonacoSettingsEditor.tsx]: It typechecks so I presume we will be notified by the compiler if it changes

**Pull request merged**